cdc/sink/codec/craft/decoder.go

```diff
@@ -0,0 +1,451 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License")
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.orglicensesLICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package craft
+
+import (
+	"encoding/binary"
+	"math"
+	"unsafe"
+
+	"github.com/pingcap/errors"
+	pmodel "github.com/pingcap/parser/model"
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/ticdc/cdc/model"
+	cerror "github.com/pingcap/ticdc/pkg/errors"
+)
+
+/// create string from byte slice without copying
+func unsafeBytesToString(b []byte) string {
+	return *(*string)(unsafe.Pointer(&b))
+}
+
+/// Primitive type decoders
+func decodeUint8(bits []byte) ([]byte, byte, error) {
+	if len(bits) < 1 {
+		return bits, 0, cerror.ErrCraftCodecInvalidData.GenWithStack("buffer underflow")
+	}
+	return bits[1:], bits[0], nil
+}
+
+func decodeVarint(bits []byte) ([]byte, int64, error) {
+	x, rd := binary.Varint(bits)
+	if rd < 0 {
+		return bits, 0, cerror.ErrCraftCodecInvalidData.GenWithStack("invalid varint data")
+	}
+	return bits[rd:], x, nil
+}
+
+func decodeUvarint(bits []byte) ([]byte, uint64, error) {
+	x, rd := binary.Uvarint(bits)
+	if rd < 0 {
+		return bits, 0, cerror.ErrCraftCodecInvalidData.GenWithStack("invalid uvarint data")
+	}
+	return bits[rd:], x, nil
+}
+
+func decodeUvarintReversed(bits []byte) (int, uint64, error) {
+	l := len(bits) - 1
+	var x uint64
+	var s uint
+	i := 0
+	for l >= 0 {
+		b := bits[l]
+		if b < 0x80 {
+			if i >= binary.MaxVarintLen64 || i == binary.MaxVarintLen64-1 && b > 1 {
+				return 0, 0, cerror.ErrCraftCodecInvalidData.GenWithStack("invalid reversed uvarint data")
+			}
+			return i + 1, x | uint64(b)<<s, nil
+		}
+		x |= uint64(b&0x7f) << s
+		s += 7
+		i++
+		l--
+	}
+	return i, x, nil
```

Comment:
It's very similar to `encoding/binary.ReadUvarint`, is it possible to use `encoding/binary.ReadUvarint` here?

Reply:
binary.ReadUvarint reads from left to right and decodeUvarintReversed reads bytes backwards. This way, the last number can be decoded without knowing where is the first byte upfront.